### ISSUE TITLE
Interest in forking

### ISSUE THREAD

**Issue by martindurant** [Pivotal-Data-Attic/pivotalrd-libhdfs3#57]
This library is critical to my library [hdfs3](https://github.com/dask/hdfs3/). No activity has occurred on this repo since the code-base's absorption back into [HAWQ](https://github.com/apache/incubator-hawq/tree/master/depends/libhdfs3), and there has been exactly [one commit](https://github.com/apache/incubator-hawq/commit/0d6a74406a4599f70b5a6cc73554d647394de382) to that location within the last year.
I am proposing to produce a new fork for libhdfs3, so that the lingering PRs can be merged, issues ironed out, and to re-enable community engagement and discussion.
Are there any thoughts on this?

cc @amyrazz44 @radarwave @kdunn-pivotal @wangzw @bdrosen96

**Comment by kdunn-pivotal** [comment]
@vvineet, @schubert, @edespino, @rvs - we *really* need to devote people to avoiding a fork, the outstanding PRs leading to this are valuable additions to libhdfs and therefor HAWQ.

@martindurant - If I recall, some of the open PRs did not have unit tests - if we can address that I'm more than willing to work on getting the PRs merged into HAWQ.

**Comment by martindurant** [comment]
Thanks for replying so quickly, @kdunn-pivotal . I wonder, why is a fork a bad idea from your point of view? In our usage, libhdfs3 is useful as an independent entity. 

If it's clean to pull fro HAWQ and compile just libhdfs3, that's a fine solution, but obviously it doesn't give the same freedom for an open development cycle. If the PRs can be merged, then I gladly withdraw that latter comment :)  I don't know how much work it would be to make new tests, especially when some functionality touched on kerberos authentication.

**Comment by schubert** [comment]
I'll defer to @rvs on this - we had discussed this when it got pulled into HAWQ  the first time - there are folks in the community (e.g. @martindurant ) who wish for it to have a separate life from HAWQ. The license can't prohibit anyone from doing so it's just a matter of does the collective group of people, HAWQ devs included, want to be a part of it (a standalone libhdfs3) or would they (HAWQ devs) like to offer an alternative like pulling libhdfs3 into a separate ASF repository for contributions?

I think last time this discussion came up that's where we stalled out.

**Comment by kdunn-pivotal** [comment]
It's definitely a very useful library all by itself; we're certainly glad the Python community has benefited.  

From my perspective, if we can keep the contributions and development work from happening in two places, every downstream consumer of libhdfs3 will win.  To be fair, libhdfs3 hasn't had substantial commit activity, as our devs have focussed on high-level integrations with Hadoop tools like Ranger.  

HAWQ is still incubating and marching towards graduation, which, I suspect will afford more flexibility in how the packaging and version control can be done... at one time a Git sub-module for it was discussed, which would allow a better dev experience and invite more contributions.

**Comment by martindurant** [comment]
I would appreciate if you could consider whether the PR and @bdrosen96 's additions (maybe others) are likely to be mergeable - perhaps in a dev/alpha branch - on the time-scale of ~days. We have a client who could well make use of the improvements. Of course, if there *were* to be a fork, any changes in it could always be patched back into HAWQ.

**Comment by wesm** [comment]
I was discussing this with @mrocklin a few weeks. Given that the HAWQ developers are working towards graduating from the Apache incubator and may not be able to spend a lot of time right now on libhdfs3 as a standalone component, I would suggest taking the approach that software vendors have taken with other Hadoop ecosystem projects and creating "downstream" releases (e.g. the versions of components in CDH or HDP are generally patched versions of the upstream Apache releases).

So the idea is that

* HAWQ is the "upstream" branch and official point of record from a governance point of view

* Developers propose patches to the upstream branch so that they can participate in the ASF community and governance process. This is very important from an IP lineage and oversight standpoint. So create HAWQ JIRAs and open pull requests or upload a patch files as a first step

* There is no particular reason to delay creating unofficial downstream builds including patches which have not yet been included in an ASF release. The packages in conda-forge (from https://github.com/conda-forge/libhdfs3-feedstock) are already unofficial, and not being voted on by the HAWQ PMC, anyway. 

I think a hard fork would be bad for the community, and should be avoided if at all possible.

**Comment by martindurant** [comment]
@wesm , glad to see you chime in.
How would one fork "downstream" only a small part of a larger project, as in this case? 

Also, I am interested in the comparison of libhdfs3 with arrow's java-linked version regarding performance and compliance to the various forms kerberos auth.

**Comment by wesm** [comment]
> How would one fork "downstream" only a small part of a larger project, as in this case?

You can make patch files from the upstream PRs and apply them to the downstream repo that contains only the component of interest

**Comment by mrocklin** [comment]
I like the idea of finding a way where the community can start making changes on libhdfs3 in a more fluid and less careful way that is still palatable to ASF concerns such as @wesm proposes.  

From a pragmatic perspective it sounds like @martindurant is trying to solve a number of active issues that he has been asked to support.  

In his position I would be tempted to fork this on github, merge in those branches, and then build and push a release on anaconda.org.  Obviously this lacks the IP hygiene provided by the ASF process, but many companies use non-ASF software happily.

So now I think this becomes a question of how to make a process that balances convenience for people like @martindurant and other fly-by contributors to this github repository on the one hand and also  keeps ASF/Pivotal happy on the other.

@wesm how much effort would your suggested process take both for the existing PRs pointed to this repository and for future PRs that show up here?  I suspect that we would lose some fraction of developers in telling them to actually submit upstream.  This fraction would be, I assume, fairly high especially for the people who tried to submit to this branch over the last year.

**Comment by wesm** [comment]
> I suspect that we would lose some fraction of developers in telling them to actually submit upstream. This fraction would be, I assume, fairly high

I would be surprised by this; these are people who took the time to write a patch and test it out in the first place. The process is not onerous:

* Create JIRA (which is basically a changelog entry explaining the purpose of your change)
* Open pull request on https://github.com/apache/incubator-hawq

I don't think this is too much to ask of contributors. The reason that this did not happen in the first place seems to be inadequate communication in this deprecated repository. 

The "downstream" repository is therefore something that you maintain as part of the packaging process (like other software vendors already do with Spark and most other Hadoop ecosystem projects) rather than a public facing repository where you advertise and solicit contributions. We aren't talking about large volumes of patches, right?

**Comment by mrocklin** [comment]
@martindurant perhaps a good first step here then is to comment on ongoing PRs pointing up to Wes's comment above, wait a few days, and then see what traction is like?  Thoughts?

**Comment by martindurant** [comment]
I think that all the people involved are pinged on here, so we can leave this up a while longer without taking any actions and see if more opinions arise.

**Comment by wesm** [comment]
I put this together as an example of what this might look like

https://github.com/wesm/libhdfs3-downstream

**Comment by mrocklin** [comment]
Well hey, that looks pretty handy!

**Comment by martindurant** [comment]
Thank you, @wesm , that's a very  useful way of structuring it. I'll get to work and push to [a copy](https://github.com/martindurant/libhdfs3-downstream) and see how well I can build libhdfs3/hdfs3 functionality from there.

**Comment by johnparke** [comment]
Thanks for forking this. I am very interested in using this library functionality.

**Comment by rvs** [comment]
@kdunn-pivotal @schubert first of all, I must say that it is very exciting to see the community interest in maintaining the native HDFS access library. That said, I'd like to bring two additional perspectives in. With my Apache Hadoop committer hat on, I must say that long term, I'd rather see community energy spent on improving the work done in https://issues.apache.org/jira/browse/HDFS-8707

The library that is part of Apache HAWQ (incubating) has a very specific set of requirements and, for example, is very unlikely to support Hadoop 3 for quite some time. In fact, this very special set of requirements is precisely why we're keeping it buried in the HAWQ's repo since we really don't want innocent bystanders to be somehow confused that it is a general purpose C/C++ access layer for accessing HDFS - it isn't -- it is a HAWQ's layer for accessing HDFS.

With that in mind, I think it is very unlikely that we, ourselves will promote this to top level or will actively accept pull requests. So fork away and who knows may be someday HAWQ itself will consider the fork (although in all likelihood HDFS-8707 is much more likely to be what HAWQ needs long term).

P.S. Do note, that there's at least 4 C/C++ HDFS access libraries I know about -- writing one to be good enough is easy. Supporting one by keeping with latest HDFS development is very, very hard.

**Comment by martindurant** [comment]
@rvs : I totally agree that it would be the best for all if HDFS-8707 were done, to provide a true reference C++ implementation on which we could build a python layer. Unfortunately, that effort has had even less work put in over the last two years than this one, which at least exposes enough of the API to be totally useful from the users' pov.
The fork being discussed here is really about providing the possibility of adding at least minor additions and fixes to the code, which has not been possible on this repo or in hawq.
You are probably right that supporting major changes to hdfs itself may require a different tack.

**Comment by rvs** [comment]
@martindurant FWIW I see a lot of activity around HDFS-8707 lately

**Comment by wesm** [comment]
Indeed, seems to be plenty of activity in the HDFS-8707 branch: https://github.com/apache/hadoop/commits/HDFS-8707
